File: ta-scheduling-webapp/src/employee_scheduling/domain.py
from .json_serialization import *

from dataclasses import dataclass
from datetime import time
from typing import Annotated
from timefold.solver.domain import *
from timefold.solver.score import HardSoftScore
from timefold.solver import SolverStatus
from pydantic import Field


class Shift(JsonDomainBase):
    id: Annotated[str, PlanningId]
    series: str
    day_of_week: str
    start_time: time
    end_time: time
    required_tas: int

class TA(JsonDomainBase):
    id: Annotated[str, PlanningId]
    name: str
    required_shifts: int
    is_grad_student: bool
    favourite_partners: Annotated[list['TA'], Field(default=None)]
    # availability: Annotated[list[Shift], Field(default_factory=list)]
    desired: Annotated[list[Shift], Field(default_factory=list)]
    undesired: Annotated[list[Shift], Field(default_factory=list)]
    unavailable: Annotated[list[Shift], Field(default_factory=list)]


# @dataclass
# class ConstraintParameters(JsonDomainBase):
#     allow_favourite_partners: Annotated[bool, ProblemFactCollectionProperty]
#     mandate_grad_undergrad: Annotated[bool, ProblemFactCollectionProperty]

@planning_entity
class ShiftAssignment(JsonDomainBase):
    id: Annotated[str, PlanningId]
    shift: Shift
    assigned_ta: Annotated[TA | None,
                        PlanningVariable,
                        Field(default=None)]

@planning_solution
class Timetable(JsonDomainBase):
    # problem facts
    shifts: Annotated[list[Shift], ProblemFactCollectionProperty]
    tas: Annotated[list[TA], ProblemFactCollectionProperty,  ValueRangeProvider]
    # constraint_parameters: Annotated[ConstraintParameters, ProblemFactProperty]
    # planning entities
    shift_assignments: Annotated[list[ShiftAssignment], PlanningEntityCollectionProperty]
    # score and solver status
    score:          Annotated[HardSoftScore | None,
                                        PlanningScore, ScoreSerializer, ScoreValidator, Field(default=None)]
    solver_status: Annotated[SolverStatus | None, Field(default=None)]
    
    
    

File: ta-scheduling-frontend/src/models/domain.ts
export enum DayOfWeek {
    Monday = "Monday",
    Tuesday = "Tuesday",
    Wednesday = "Wednesday",
    Thursday = "Thursday",
    Friday = "Friday",
    Saturday = "Saturday",
    Sunday = "Sunday"
}

export class Shift {
    id: string
    series: string
    dayOfWeek: DayOfWeek
    startTime: string
    endTime: string
    requiredTas: number
    // Optional
    alias: string = "DEFAULT"
    shiftDate: string = "1900-01-01"

    constructor(id: string, series: string, day_of_week: DayOfWeek, start_time: string, end_time: string, required_tas: number, alias: string = "DEFAULT", shift_date: string = "1900-01-01") {
        this.id = id
        this.series = series
        this.dayOfWeek = day_of_week
        this.startTime = start_time
        this.endTime = end_time
        this.requiredTas = required_tas
        this.alias = alias
        this.shiftDate = shift_date
    }
}

export class TA {
    id: string
    name: string
    requiredShifts: number
    desired: Shift[] = []
    undesired: Shift[] = []
    unavailable: Shift[] = []
    // favourite_partners: TA[] = None
    isGradStudent: boolean = true

    constructor(id: string, name: string, required_shifts: number, desired: Shift[] = [], undesired: Shift[] = [], unavailable: Shift[] = [], is_grad_student: boolean = true) {
        this.id = id
        this.name = name
        this.requiredShifts = required_shifts
        this.desired = desired
        this.undesired = undesired
        this.unavailable = unavailable
        this.isGradStudent = is_grad_student
    }


}

export function get_status_for_shift(ta: TA, shift: Shift): string {
    if (ta.desired.map(shift => shift.id).includes(shift.id))
        return 'Desired'
    if (ta.undesired.map(shift => shift.id).includes(shift.id))
        return 'Undesired'
    if (ta.unavailable.map(shift => shift.id).includes(shift.id))
        return 'Unavailable'
    return 'Neutral'
}

export class ShiftAssignment {
    id: string
    assignedTa?: TA
    shift: Shift

    constructor(id: string, shift: Shift, assigned_ta?: TA) {
        this.id = id
        this.assignedTa = assigned_ta
        this.shift = shift
    }
}

export class Score {
    hardScore: number
    softScore: number
    mediumScore: number

    constructor(hardScore: number = 0, softScore: number = 0, mediumScore: number = 0) {
        this.hardScore = hardScore
        this.softScore = softScore
        this.mediumScore = mediumScore
    }
}

export class Timetable {
    id: string
    // problem facts
    shifts: Shift[]
    tas: TA[]
    // planning entities
    shiftAssignments: ShiftAssignment[]
    score: Score

    constructor(id: string, shifts: Shift[], tas: TA[], shift_assignments: ShiftAssignment[], score: Score = new Score()) {
        this.id = id
        this.shifts = shifts
        this.tas = tas
        this.shiftAssignments = shift_assignments
        this.score = score
    }
}
